src/analysis/signature_help.ml

```diff
@@ -65,74 +70,93 @@ let pp_parameter env label ppf ty =
     Format.fprintf ppf "?%s:%a" l (pp_parameter_type env) (unwrap_option ty)
 
 (* record buffer offsets to be able to underline parameter types *)
-let print_parameter_offset buffer env label ty =
+let print_parameter_offset ?arg:argument buffer env label ty =
   let ppf = Format.formatter_of_buffer buffer in
   let param_start = Buffer.length buffer in
   Format.fprintf ppf "%a%!" (pp_parameter env label) ty;
   let param_end = Buffer.length buffer in
   Format.pp_print_string ppf " -> ";
   Format.pp_print_flush ppf ();
-  (param_start, param_end)
+  { label; param_start; param_end; argument }
 
-let application_signature = function
-  (* provide signature information for applied functions *)
-  | (_, Expression earg) :: (_, Expression { exp_desc =
-      Texp_apply ({ exp_type = { desc = Tarrow _; _ }; _ } as efun, eargs); _ }) :: _ ->
-    Type_utils.Printtyp.reset ();
-    let buffer = Buffer.create 16 in
-    let rec separate_signature ?(i=0) ?(param_offsets=[]) ?active_param args ty =
-      match (args, ty) with
-      | [], { Types.desc = Tarrow (label, ty1, ty2, _) } ->
-        let offset = print_parameter_offset buffer efun.exp_env label ty1 in
-        let param_offsets = offset::param_offsets in
-        separate_signature args ty2 ~i:(succ i) ~param_offsets ?active_param
-
-      | arg :: args, { Types.desc = Tarrow (label, ty1, ty2, _) } ->
-        let offset = print_parameter_offset buffer efun.exp_env label ty1 in
-        let param_offsets = offset::param_offsets in
-        let active_param = match arg with
-          | (l, Some e) when l = label && e == earg -> Some i
-          | _ -> active_param
-        in
-        separate_signature args ty2 ~i:(succ i) ~param_offsets ?active_param
-
-      (* end of function type, print remaining type without recording offsets *)
+let separate_function_signature ?args (e : Typedtree.expression) =
```

Comment:
I see that this function is only called twice: with & without the optional argument. To me that hints that the optionality of this variable doesn't buy us much. I usually prefer to avoid optional variables in such situations, as it seems like the right thing to do is to make the caller think about whether `args` should be passed.